### PR TITLE
Cas 9019 Align old lattice stats with sicmath stats framework for computing stats of complex-values

### PR DESCRIPTION
**Pull request by dmehring**
Changes include:
* explicitly implement +/- operators for heterogenous operand types (Complex, DComplex) in Complex.h
* Add another StatisticsUtilities::accumulate() variation to accommodate StatsTiledCollapser accumulations
* Replace LatticeStatsSpeicialize calls with StatisticsUtilities calls in StatsTiledCollapser
* reformat StatsTiledCollapser by replacing tabs with spaces
* refactor StatsTiledCollapser so that there is only one conditional branch for both include and exclude ranges, rather than the previous two
* increase test coverage in tLatticeStatistics to more fully explore parameter space for complex valued lattices and to ensure that old StatsTiledCollapser and new scimath stats framework produce the identical results for complex-valued lattices.

### CODE REVIEW

File: lattices/LatticeMath/StatsTiledCollapser.tcc [hunk 1]
```diff
@@ -47,7 +49,7 @@ template <class T, class U>
 void StatsTiledCollapser<T,U>::initAccumulator (uInt n1, uInt n3) {
    _sum = new Block<U>(n1*n3);
    _sumSq = new Block<U>(n1*n3);
-   _npts = new Block<U>(n1*n3);
+   _npts = new Block<Double>(n1*n3);
```
Comment (juliantaylor) [inline lattices/LatticeMath/StatsTiledCollapser.tcc]: are n1 and n3 data dimensions? if so care should be taken to not overflow the int, e.g. by using int64

File: lattices/LatticeMath/StatsTiledCollapser.tcc [hunk 2]
```diff
@@ -71,166 +73,138 @@ void StatsTiledCollapser<T,U>::initAccumulator (uInt n1, uInt n3) {
...
-	T& dataMin = (*_min)[index];
-	T& dataMax = (*_max)[index];
-	U& mean = (*_mean)[index];
-	U& variance = (*_variance)[index];
-	U& nvariance = (*_nvariance)[index];
-	Bool& minMaxInit = (*_initMinMax)[index];
+    // Process the data in the current chunk.   Everything in this
+    // chunk belongs in one output location in the storage
+    // lattices
+    uInt index = index1 + index3*_n1;
```
Comment (juliantaylor) [inline lattices/LatticeMath/StatsTiledCollapser.tcc]: possibly same overflow problem

File: lattices/LatticeMath/StatsTiledCollapser.tcc [hunk 3]
```diff
@@ -239,18 +213,21 @@ void StatsTiledCollapser<T,U>::endAccumulator(
     Bool deleteRes;
     U* res = result.getStorage (deleteRes);
     U* resptr = res;
-
     U* sumPtr = _sum->storage();
     U* sumSqPtr = _sumSq->storage();
-    U* nPtsPtr = _npts->storage();
+    CountedPtr<Block<DComplex> > nptsComplex;
+    if (! isReal(whatType(resptr))) {
+        nptsComplex = new Block<DComplex>(_n1*_n3);
```
Comment (juliantaylor) [inline lattices/LatticeMath/StatsTiledCollapser.tcc]: overflow?

File: lattices/LatticeMath/StatsTiledCollapser.tcc [hunk 4]
```diff
@@ -239,18 +213,21 @@ void StatsTiledCollapser<T,U>::endAccumulator(
...
     U* meanPtr = _mean->storage();
     U* variancePtr = _variance->storage();
     const T* minPtr = _min->storage();
     const T* maxPtr = _max->storage();
-
     uInt i,j;
     U* resptr_root = resptr;
-    for (i=0; i<_n3; i++) {
+    for (i=0; i<_n3; ++i) {
        resptr = resptr_root + (Int(LatticeStatsBase::NPTS) * _n1);
```
Comment (juliantaylor) [inline lattices/LatticeMath/StatsTiledCollapser.tcc]: overflow?

File: lattices/LatticeMath/StatsTiledCollapser.tcc [hunk 5]
```diff
@@ -283,18 +260,29 @@ void StatsTiledCollapser<T,U>::endAccumulator(
...
 
-    result.putStorage (res, deleteRes);
+template <class T, class U>
+void StatsTiledCollapser<T,U>::_convertNPts(
+    DComplex*& nptsPtr, CountedPtr<Block<Double> > npts,
+    CountedPtr<Block<DComplex> > nptsComplex
+) const {
+    DComplex* storage = nptsComplex->storage();
+    Double* realStorage = npts->storage();
+    for (uInt i=0; i<_n1*_n3; ++i) {
```
Comment (juliantaylor) [inline lattices/LatticeMath/StatsTiledCollapser.tcc]: overflow?

**Comment by dmehring** [comment]
Thanks. Changed the uInt's to uInt64's.

**Comment by dmehring** [comment]
It looks like there is a problem with the continuous integration system and not with my code. Is that correct?

**Comment by juliantaylor** [comment]
yes the macs are broken again, don't worry about those

I don't think keeping initAccumulator(uInt, uInt) is very useful, uInt is implicitly convertable to uInt64 so source compatibility is retained.
Removing it breaks binary compatibility but having it error out on runtime is even worse.
As I understand it packagers have now accepted to bump binary compatibility version on each release so breaking it is not such a big deal anymore.

**Comment by dmehring** [comment]
Works for me. It is difficult to keep up with the rules of what can be changed and what cannot...

**Comment by juliantaylor** [comment]
thanks

**Pull request merged**